# Pandas DataFrame - Basic Properties

In [1]:
# importing the libraries
import numpy as np
import pandas as pd

In [2]:
# reading the csv file
df = pd.read_csv('../data/03-Pandas/tips.csv')

In [4]:
# listing the elements inside the data frame
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  
0    Christy Cunningham  3560325168603410    Sun2959  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
4          Tonya Carter  4832732618637221    Sun2251  
..                  ...               ...        ...  
239       Michael Avila  5296068606052842    Sat2657  
240      Monica Sanders  3506806155565404    Sat1766  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  
243     Michelle Hardin  3511451626698139    Thur672  

[244 rows x 11 columns]

Până în acest moment am învățat cum anume să citim date dintr-un fișier csv. Având aceste date salvate într-un data frame o să începem să explicăm ce anume se există în acest data frame după care o să trecem prin câteva metode și atribute generale pentru un data frame pentru a putea înțelege mai bine ce anume se găsește în cadrul acestui data frame. 

Acest data frame reprezintă un set de date unde un ospătar a înregistrat informații despre fiecare tip pe care l-a primit pe o perioadă de câteva luni de zile. După cum se observă, în cadrul acestui data frame există un număt de 11 coloane și 244 de rânduri. Aceste coloane conțin informații referitoare la toatul notei de plată, la bacșișul care s-a lăsat, la numărul total de persoane care au fost la masă, sexul persoanei care a plătit și alte informații. Un prim lucru pe care poate dorim să îl aflăm despre acest data frame reprezintă numele coloanelor. Pentru a returna toate denumirile coloanelor putem să ne utilizăm de atributul 'columns'. Acest atribut returnează sub formă de listă toate numele coloanelor

In [5]:
df.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size',
       'price_per_person', 'Payer Name', 'CC Number', 'Payment ID'],
      dtype='object')

Numele coloanelor sunt importante deoarece dacă dorim să extragem anumite informații din cadrul unuei singure coloane atunci trebuie să îi oferim numele acelei coloane. De asemena, un atribut destul de important ce îl putem utiliza este 'index'. Acest atribut ne returnează valorile pentru fiecare index din cadrul unui data frame

In [6]:
df.index

RangeIndex(start=0, stop=244, step=1)

În cazul de față putem să observăm că acest index reprezintă un range creat automat cu valori de la 0 la 244. Dacă acest data frame avea un index de altă formă (un string label), atunci valorile acestui index ar fi fost returnate sub forma unuei liste precum ceea ce a fost returnat de către atributul columns

Pentru acest data frame avem un număr de 244 de coloane. Dacă dorim să afișăm doar primele rânduri din cadrul unui data frame, atunci putem utiliza metoda 'head()'. Metoda respectivă ne returnează primele 5 rânduri din cadrul unui data frame

In [7]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

În mod default, metoda 'head()' ne returnează primele 5 rânduri dintr-un data frame. Dacă dorim, în cadrul parantezelor putem să trecem o valoare care o să reprezinte numărul de rânduri care dorim să fie afișat. Dacă dorim să ne afișeze doar primele 2 rânduri din data frame, atunci pentru metoda 'head()' o să îi oferim ca și argument valorea 2

In [8]:
df.head(2)

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608

De asemenea există și o metodă prin care putem extrage ultimele valori din cadrul unui data frame, iar acea metodă poartă denumirea de 'tail()'. Metoda respectivă are același comportament precum 'head()', ne retunrează 5 rânduri dacă nu se oferă nicio valoare ca și argument

In [9]:
df.tail()

total_bill   tip     sex smoker   day    time  size  price_per_person  \
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

          Payer Name         CC Number Payment ID  
239    Michael Avila  5296068606052842    Sat2657  
240   Monica Sanders  3506806155565404    Sat1766  
241       Keith Wong  6011891618747196    Sat3880  
242     Dennis Dixon     4375220550950      Sat17  
243  Michelle Hardin  3511451626698139    Thur672

După cum am văzut și anterior, pentru a afla informații generale despre un data frame putem să utilizăm metoda 'info()'. Această metodă ne afișează date generale despre data frame, cum ar fi numărul total de rânduri și de coloane, denumirea fiecărei coloane, tipul de date stocat în cadru fiecărui coloană (de notat faptul că pandas în loc de string trece ca și denumire object. Prin urmare, unde avem object înseamnă că avem de aface cu un tip de date de tipul string)

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   total_bill        244 non-null    float64
 1   tip               244 non-null    float64
 2   sex               244 non-null    object 
 3   smoker            244 non-null    object 
 4   day               244 non-null    object 
 5   time              244 non-null    object 
 6   size              244 non-null    int64  
 7   price_per_person  244 non-null    float64
 8   Payer Name        244 non-null    object 
 9   CC Number         244 non-null    int64  
 10  Payment ID        244 non-null    object 
dtypes: float64(3), int64(2), object(6)
memory usage: 21.1+ KB


După cum se poate observa, avem de a face și cu unele coloane care rețin date de tipul integer sau float. Pentru aceste date putem să executăm anumite funcții care să ne returneze valoare medie din cadrul coloanei, valorea maximă, minimă, deviația standard sau alte valori. Pentru a putea calcula și vizualiza toate aceste date, pandas ne pune la dispoziție o metodă denumită 'descibe()' care face automat acesta calcule și le afișează

In [12]:
df.describe()

total_bill         tip        size  price_per_person     CC Number
count  244.000000  244.000000  244.000000        244.000000  2.440000e+02
mean    19.785943    2.998279    2.569672          7.888197  2.563496e+15
std      8.902412    1.383638    0.951100          2.914234  2.369340e+15
min      3.070000    1.000000    1.000000          2.880000  6.040679e+10
25%     13.347500    2.000000    2.000000          5.800000  3.040731e+13
50%     17.795000    2.900000    2.000000          7.255000  3.525318e+15
75%     24.127500    3.562500    3.000000          9.390000  4.553675e+15
max     50.810000   10.000000    6.000000         20.270000  6.596454e+15

Dacă dorim să inversăm datele din acel data frame (ca denumirea coloanelor să fie trecută ca și index, respectiv invers) în cazul în care ni se pare că este mai ușor să vizualizăm datele în acel format, atunci putem să utilizăm metoda 'transpose()' care o să realizeze acest lucru

In [13]:
df.describe().transpose()

count          mean           std           min  \
total_bill        244.0  1.978594e+01  8.902412e+00  3.070000e+00   
tip               244.0  2.998279e+00  1.383638e+00  1.000000e+00   
size              244.0  2.569672e+00  9.510998e-01  1.000000e+00   
price_per_person  244.0  7.888197e+00  2.914234e+00  2.880000e+00   
CC Number         244.0  2.563496e+15  2.369340e+15  6.040679e+10   

                           25%           50%           75%           max  
total_bill        1.334750e+01  1.779500e+01  2.412750e+01  5.081000e+01  
tip               2.000000e+00  2.900000e+00  3.562500e+00  1.000000e+01  
size              2.000000e+00  2.000000e+00  3.000000e+00  6.000000e+00  
price_per_person  5.800000e+00  7.255000e+00  9.390000e+00  2.027000e+01  
CC Number         3.040731e+13  3.525318e+15  4.553675e+15  6.596454e+15

## Recapitulare

În acest tutorial am învățat următoarele lucruri:

    1. Cum să afișăm toate denumirle coloanelor sub forma unei liste de valori

        df.columns

            # aceste valori o să ne folosească atunci când dorim să extragem anumite date dintr-o coloană
    
    2. Cum să accesăm valorile index-ului dintr-un data frame sub forma unei liste de valori

        df.index

    3. Cum să afișăm un număr de rânduri dintr-un data frame fie de la început sau de la final

        df.head()

        df.head(2)

        df.tail()

        df.tail(2)

    4. Cum să afișăm informațiile generale despre un data frame

        df.info()

    5. În pandas, un string este trecut ca și fiind object (sub denumirea de object)

    6. Cum putem afla informații generale despre coloanele care dețin informații numerice

        df.describe()

        df.describe().transpose()

            # metoda transpose de la final ne inversează coloanele cu rândurile pentru o astfel de vizualizare a datelor